src/allmydata/util/eliotutil.py

```diff
@@ -320,7 +332,15 @@ def decorate_log_call_deferred(f):
         def logged_f(*a, **kw):
             # Use the action's context method to avoid ending the action when
             # the `with` block ends.
-            with start_action(action_type=action_type).context():
+            kwargs = {k: bytes_to_unicode(True, kw[k]) for k in kw}
+            # Remove complex (unserializable) objects from positional args to
+            # prevent eliot from throwing errors when it attempts serialization
+            args = {
+                "arg_" + str(pos): bytes_to_unicode(True, a[pos])
```

Comment:
My thinking is: `args` are usually a list .. so suddenly seeing them as a `dict` with `"arg_0": ...` etc is .. weird. AFAIK, `repr()` always returns a string, which is always JSON-serializable. So, you should be able to simply map the `args` list into another list that doesn't contain anything except numbers, strings, lists or dicts (because if a thing _is_ in-JSON-serializable, you just `repr()` it first, then it is).

Reply:
It also might be better to have a `serialize_as_json()` method instead of checking, then doing some sort of serialization (although as already mentioned, I don't think `bytes_to_unicode` is the right thing there). So then `serialize_as_json` just tried to serialize, and if that fails, serializes the `repr()`?

(My reasoning for suggesting `repr` here at all is because this is for logging .. so if you're logging the args you likely want to see _something_ for every arg instead of only seeing args that happen to also be JSON-able).

Reply:
I don't feel great about all of this automatic JSON and repr()'ing.  This use of Eliot doesn't seem like it plays to the strengths of Eliot.

It's also not as if this is the only way to do Eliot logging.  If there is an argument that is not JSON encodeable and a developer wants it in the logs, maybe they should stop using `log_call_deferred` and create a custom field, message, or action type that can do the serialization.